src/signalrclient/messagepack_hub_protocol.cpp

```
@@ -0,0 +1,427 @@
+#include "binary_message_formatter.h"
+
+namespace signalr
+{
+    // replaces use of msgpack::sbuffer when packing so that we can write directly to the string instead of writing to another buffer and copying to a string later
+    class string_wrapper
+    {
+    public:
+        std::string str;
+
+        string_wrapper& write(const char* buf, size_t s)
+        {
+            str.append(buf, s);
+            return *this;
+        }
+    };
+
+    signalr::value createValue(const msgpack::object& v)
+    {
+        switch (v.type)
+        {
+        case msgpack::type::object_type::BOOLEAN:
+            return signalr::value(v.via.boolean);
+        case msgpack::type::object_type::FLOAT64:
+        case msgpack::type::object_type::FLOAT32:
+            return signalr::value(v.via.f64);
+        case msgpack::type::object_type::POSITIVE_INTEGER:
+            return signalr::value((double)v.via.u64);
+        case msgpack::type::object_type::NEGATIVE_INTEGER:
+            return signalr::value((double)v.via.i64);
+        case msgpack::type::object_type::STR:
+            return signalr::value(v.via.str.ptr, v.via.str.size);
+        case msgpack::type::object_type::ARRAY:
+        {
+            std::vector<signalr::value> vec;
+            for (size_t i = 0; i < v.via.array.size; ++i)
+            {
+                vec.push_back(createValue(*(v.via.array.ptr + i)));
+            }
+            return signalr::value(std::move(vec));
+        }
+        case msgpack::type::object_type::MAP:
+        {
+            std::map<std::string, signalr::value> map;
+            for (size_t i = 0; i < v.via.map.size; ++i)
+            {
+                auto key = (v.via.map.ptr + i)->key.as<std::string>();
+                map.insert({ key, createValue((v.via.map.ptr + i)->val) });
+            }
+            return signalr::value(std::move(map));
+        }
+        case msgpack::type::object_type::NIL:
+        case msgpack::type::object_type::BIN:
```

We should probably throw for unsupported types, and add todo for binary.